### PR TITLE
CORE-16202 Multi-Source Event Mediator - Factories for Flow Engine and Flow Mapper

### PR DESCRIPTION
**Pull request by mbrkic-r3**
Multi-Source Event Mediator factories that create Multi-Source Event Mediator for Flow Engine and Flow Mapper. Factories set up all needed consumers, messaging clients and message routing.

### CODE REVIEW

File: components/flow/flow-mapper-service/src/test/kotlin/net/corda/session/mapper/service/messaging/mediator/FlowMapperEventMediatorFactoryImplTest.kt [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+import org.junit.jupiter.api.BeforeEach
+import org.junit.jupiter.api.Test
+import org.mockito.Mockito.`when`
+import org.mockito.kotlin.any
+import org.mockito.kotlin.mock
+
+class FlowMapperEventMediatorFactoryImplTest {
+    private lateinit var flowMapperEventMediatorFactory: FlowMapperEventMediatorFactory
+    private val flowMapperEventExecutorFactory = mock<FlowMapperEventExecutorFactory>()
+    private val mediatorConsumerFactoryFactory = mock<MediatorConsumerFactoryFactory>()
```
Comment (emilybowe) [inline components/flow/flow-mapper-service/src/test/kotlin/net/corda/session/mapper/service/messaging/mediator/FlowMapperEventMediatorFactoryImplTest.kt:22]: nit--> I think FactoryFactory is not ideal but it is pretty clear, I've left some ideas if we want to go with a different name if you like any of these:
mediatorConsumerFactoryCreator
mediatorConsumerFactoryProvider
mediatorConsumerMetaFactory
mediatorConsumerPreFactory

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/messaging/mediator/FlowEventMediatorFactory.kt [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+import net.corda.messaging.api.mediator.MultiSourceEventMediator
+
+/**
+ * Creates a Multi-Source Event Mediator for flow engine.
+ */
+interface FlowEventMediatorFactory {
+    /**
+     * Creates a Multi-Source Event Mediator for flow engine.
+     *
+     * @param configs Map of configurations.
```
Comment (emilybowe) [inline components/flow/flow-service/src/main/kotlin/net/corda/flow/messaging/mediator/FlowEventMediatorFactory.kt]: As it's a map I'd mention what the key is

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/messaging/mediator/FlowEventMediatorFactoryImpl.kt [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+                else -> {
+                    val eventType = event?.let { it::class.java }
+                    throw IllegalStateException("No route defined for event type [$eventType]")
+                }
+            }
+        }
+    }
+
+    /**
+     * Deserialized message payload if it is a [ByteArray] (seems to be the case for external events).
```
Comment (emilybowe) [inline components/flow/flow-service/src/main/kotlin/net/corda/flow/messaging/mediator/FlowEventMediatorFactoryImpl.kt]: Should this be _Deserialized --> Deserialize_ ?

**Review by emilybowe — DISMISSED** [review]
LGTM, left few small comments 👍

File: components/flow/flow-mapper-service/src/test/kotlin/net/corda/session/mapper/service/messaging/mediator/FlowMapperEventMediatorFactoryImplTest.kt (hunk 1 above)
Comment (mbrkic-r3) [reply components/flow/flow-mapper-service/src/test/kotlin/net/corda/session/mapper/service/messaging/mediator/FlowMapperEventMediatorFactoryImplTest.kt:22]: [My opinion](https://github.com/corda/corda-runtime-os/pull/4694#discussion_r1335717690) on this is:

> I don't think having FactoryFactory is a bad thing, it's consistent naming where you remove the last Factory suffix and know exactly what that factory is creating.

Until we have naming convention for this it's best to be consistent, and we already have [EntityManagerFactoryFactory](https://github.com/corda/corda-runtime-os/blob/975a62e1bb917dba7f7a7ad4b2513dea45ecb142/libs/db/db-orm/src/main/kotlin/net/corda/orm/EntityManagerFactoryFactory.kt#L8)

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/messaging/mediator/FlowEventMediatorFactory.kt (hunk 2 above)
Comment (mbrkic-r3) [reply components/flow/flow-service/src/main/kotlin/net/corda/flow/messaging/mediator/FlowEventMediatorFactory.kt]: Fixed

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/messaging/mediator/FlowEventMediatorFactoryImpl.kt (hunk 3 above)
Comment (mbrkic-r3) [reply components/flow/flow-service/src/main/kotlin/net/corda/flow/messaging/mediator/FlowEventMediatorFactoryImpl.kt]: Fixed

**Review by emilybowe — APPROVED** [review]
LGTM

**Pull request merged**